src/core/search/compressed_list.cc

```diff
@@ -0,0 +1,210 @@
+}
+
+bool operator==(const CompressedList::Iterator& l, const CompressedList::Iterator& r) {
+  return l.diffs_.data() == r.diffs_.data() && l.diffs_.size() == r.diffs_.size();
+}
+
+bool operator!=(const CompressedList::Iterator& l, const CompressedList::Iterator& r) {
+  return !(l == r);
+}
+
+void CompressedList::Iterator::ReadNext() {
+  if (diffs_.empty()) {
+    diffs_ = {nullptr, 0};
+    return;
+  }
+
+  uint32_t base = stash_.value_or(0);
+  auto [diff, read] = CompressedList::ReadVarLen(diffs_);
+
+  stash_ = base + diff;
+  diffs_.remove_prefix(read);
+}
+
+CompressedList::Iterator CompressedList::begin() const {
+  return Iterator{*this};
+}
+
+CompressedList::Iterator CompressedList::end() const {
+  return Iterator{};
+}
+
+CompressedList::SortedBackInserter::SortedBackInserter(CompressedList* list)
+    : last_{0}, list_{list} {
+}
+
+CompressedList::SortedBackInserter& CompressedList::SortedBackInserter::operator=(uint32_t value) {
+  DCHECK_LE(last_, value);
+  if (value > last_) {
+    list_->PushBackDiff(value - last_);
+    last_ = value;
+  }
+  return *this;
+}
+
+// Simply encode difference and add to end of diffs array
+void CompressedList::PushBackDiff(uint32_t diff) {
+  array<uint8_t, 16> buf;
+  auto diff_span = WriteVarLen(diff, absl::MakeSpan(buf));
+  diffs_.insert(diffs_.end(), diff_span.begin(), diff_span.end());
+}
+
+// Do a linear scan by encoding all diffs to find value
+CompressedList::EntryLocation CompressedList::LowerBound(uint32_t value) const {
+  uint32_t entry = 0, prev_entry = 0;
+  absl::Span<const uint8_t> diffs_tail{diffs_}, entry_span{};
+
+  while (entry < value && !diffs_tail.empty()) {
+    auto [diff, read] = ReadVarLen(diffs_tail);
+
+    prev_entry = entry;
+    entry += diff;
+
+    entry_span = diffs_tail.subspan(0, read);
+    diffs_tail.remove_prefix(read);
+  }
+
+  return EntryLocation{entry, prev_entry, entry_span};
+}
+
+// Insert has linear complexity. It tries to find between which two entries A and B the new value V
+// needs to be inserted. Then it computes the differences dif1 = V - A and diff2 = B - V that need
+// to be stored to encode the triple A V B. Those are stored where diff0 = B - A was previously
+// stored, possibly extending the vector
+void CompressedList::Insert(uint32_t value) {
+  auto bound = LowerBound(value);
+
+  // At least one element was read and it's equal to value: return to avoid duplicate
+  if (bound.value == value && !bound.diff_span.empty())
+    return;
+
+  // We're inserting below unconditionally
+  size_++;
+
+  // Value is bigger than any other (or list is empty): append required diff at the end
+  if (value > bound.value || bound.diff_span.empty()) {
+    PushBackDiff(value - bound.value);
+    return;
+  }
+
+  // Now the list certainly contains the bound B > V and possibly A < V (or 0 by default),
+  // so we need to encode both differences diff1 and diff2
+  DCHECK_GT(bound.value, value);
+  DCHECK_LT(bound.prev_value, value);
```

Comment:
Should this be `DCHECK_LE()` in the case both are `0`?

Reply:
Good catch. Fixed and added a testcase